README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# itp

[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/paulnorthrop/itp?branch=main&svg=true)](https://ci.appveyor.com/project/paulnorthrop/itp)
[![R-CMD-check](https://github.com/paulnorthrop/itp/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/paulnorthrop/itp/actions/workflows/R-CMD-check.yaml)
[![Coverage
Status](https://codecov.io/github/paulnorthrop/itp/coverage.svg?branch=main)](https://app.codecov.io/github/paulnorthrop/itp?branch=main)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/itp)](https://cran.r-project.org/package=itp)
[![Downloads
(monthly)](https://cranlogs.r-pkg.org/badges/itp?color=brightgreen)](https://cran.r-project.org/package=itp)
[![Downloads
(total)](https://cranlogs.r-pkg.org/badges/grand-total/itp?color=brightgreen)](https://cran.r-project.org/package=itp)

## The Interpolate, Truncate, Project (ITP) Root-Finding Algorithm

The **itp** package implements the Interpolate, Truncate, Project (ITP)
root-finding algorithm of [Oliveira and Takahashi
(2021)](https://doi.org/10.1145/3423597). Each iteration of the
algorithm results in a bracketing interval for the root that is narrower
than the previous interval. It’s performance compares favourably with
existing methods on both well-behaved functions and ill-behaved
functions while retaining the worst-case reliability of the bisection
method. For details see the authors’ [Kudos
summary](https://www.growkudos.com/publications/10.1145%25252F3423597/reader)
and the Wikipedia article [ITP
method](https://en.wikipedia.org/wiki/ITP_method).

### Examples

We use three examples from Section 3 of [Oliveira and Takahashi
(2021)](https://doi.org/10.1145/3423597) to illustrate the use of the
`itp` function. Each of these functions has a root in the interval
$(-1, 1)$. The function can be supplied either as an R function or as an
external pointer to a C++ function.

``` r
library(itp)
```

#### A continuous function

The Lambert function $l(x) = xe^x - 1$ is continuous.

<img src="man/figures/README-lambert-1.png" style="display: block; margin: auto;" />

The `itp` function finds an estimate of the root, that is, $x^*$ for
which $f(x^*)$ is (approximately) equal to 0. The algorithm continues
until the length of the interval that brackets the root is smaller than
$2 \epsilon$, where $\epsilon$ is a user-supplied tolerance. The default
is $\epsilon = 10^{-10}$.

First, we supply an R function that evaluates the Lambert function.

``` r
# Lambert, using an R function
lambert <- function(x) x * exp(x) - 1
itp(lambert, c(-1, 1))
#> function: lambert 
#>       root     f(root)  iterations  
#>     0.5671   2.048e-12           8
```

Now, we create an external pointer to a C++ function that has been
provided in the `itp` package and pass this pointer to the function
`itp()`. For more information see the [Overview of the itp
package](https://paulnorthrop.github.io/itp/articles/itp-vignette.html)
vignette.

``` r
# Lambert, using an external pointer to a C++ function
lambert_ptr <- xptr_create("lambert")
itp(lambert_ptr, c(-1, 1))
#> function: lambert_ptr 
#>       root     f(root)  iterations  
#>     0.5671   2.048e-12           8
```

#### The function `itp_c`

Also provided is the function `itp_c`, which is equivalent to `itp`, but
the calculations are performed entirely using C++, and the arguments
differ slightly: `itp_c` has a named required argument `pars` rather
than `...` and it does not have the arguments `interval`, `f.a` or
`f.b`.

``` r
# Calling itp_c()
res <- itp_c(lambert_ptr, pars = list(), a = -1, b = 1)
res
#> function:  
#>       root     f(root)  iterations  
#>     0.5671   2.048e-12           8
```

#### A discontinuous function

The staircase function $s(x) = \lceil 10 x - 1 \rceil + 1/2$ is
discontinuous.

<img src="man/figures/README-staircase-1.png" style="display: block; margin: auto;" />

The `itp` function finds the discontinuity at $x = 0$ at which the sign
of the function changes. The value of 0.5 returned for the root
`res$root` is the midpoint of the bracketing interval `[res$a, res$b]`
at convergence.

``` r
# Staircase
staircase <- function(x) ceiling(10 * x - 1) + 1 / 2
res <- itp(staircase, c(-1, 1))
print(res, all = TRUE)
#> function: staircase 
#>       root     f(root)  iterations           a           b         f.a  
#>  7.404e-11         0.5          31           0   1.481e-10        -0.5  
#>        f.b   precision  
#>        0.5   7.404e-11
```

#### A function with multiple roots

The Warsaw function
$w(x) = I(x > -1)\left(1 + \sin\left(\frac{1}{1+x}\right)\right)-1$ has
multiple roots.

<img src="man/figures/README-warsaw-1.png" style="display: block; margin: auto;" />

When the initial interval is $[-1, 1]$ the `itp` function finds the root
$x \approx -0.6817$. There are other roots that could be found from a
different initial interval.

``` r
# Warsaw
warsaw <- function(x) ifelse(x > -1, sin(1 / (x + 1)), -1)
itp(warsaw, c(-1, 1))
#> function: warsaw 
#>       root     f(root)  iterations  
#>    -0.6817  -5.472e-11          11
```

### Installation

To get the current released version from CRAN:

``` r
install.packages("itp")
```

### Vignette

See the [Overview of the itp
package](https://paulnorthrop.github.io/itp/articles/itp-vignette.html)
vignette, which can also be accessed using
`vignette("itp-vignette", package = "itp")`.


NEWS.md
# itp 1.2.1

## Bug fixes and minor improvements

* The issue described at https://github.com/RcppCore/Rcpp/issues/1287 has been fixed to avoid WARNINGs from CRAN checks on some platforms. Thank you to Dirk Eddelbuettel for providing the fix so quickly!

# itp 1.2.0

## New features

* The function `itp_c`, which implements the ITP algorithm entirely in C++.

# itp 1.1.0

## New features

* The argument `f` to the function `itp()` can now be either an R function or an external pointer to a C++ function. In the latter case, `itp()` makes use of the [Passing user-supplied C++ functions](https://gallery.rcpp.org/articles/passing-cpp-function-pointers/) framework offered by the Rcpp package. For the simple examples given in the `itp` package only a modest improvement in speed is observed (and expected).  However, being able to call `itp()` on the C++ side may have benefits in more challenging problems.  
* The package vignette and README file include examples based on external pointers to a C++ function.
* A plot method is provided, which plots the function `f` provided to `itp` over the input interval `(a, b)`.

## Bug fixes and minor improvements

* Explicit examples of passing arguments to the input function `f` have been added.

# itp 1.0.1

## Bug fixes and minor improvements

* A bug in the part of `itp()` in which the bracketing interval has been fixed. It caused incorrect results in cases where the function is locally decreasing.
* Tests have been added to check the correction of this bug.


tests/testthat.R
library(testthat)
library(itp)

test_check("itp")


tests/testthat/test-plot.R
# Check that plot.itp works for
# (a) R function with no additional arguments
# (b) C++ pointer with no additional arguments
# (c) R function with additional arguments
# (d) C++ pointer with additional arguments

## Lambert

# (a)
lambert <- function(x) x * exp(x) - 1
res <- itp(lambert, c(-1, 1))
test_that("Lambert in R: plot.itp OK", {
  testthat::expect_silent(plot(res))
})

# (b)
lambert_ptr <- xptr_create("lambert")
res <- itp(lambert_ptr, c(-1, 1))
test_that("Lambert in C++: plot.itp OK", {
  testthat::expect_silent(plot(res))
})

## Trigonometric 1

# (c)
trig1 <- function(x, root) tan(x - root)
res <- itp(trig1, c(-1, 1), root = 1 / 10)
test_that("Trig 1 in R: plot.itp OK", {
  testthat::expect_silent(plot(res))
})

# (d)
trig1_ptr <- xptr_create("trig1")
itp(f = trig1_ptr, c(-1, 1), root = 1 / 10)
test_that("Lambert in C++: plot.itp OK", {
  testthat::expect_silent(plot(res))
})


tests/testthat/test-print.R
# Check that the print function works

# The example used in the Wikipedia entry for the ITP method
wiki <- function(x) x ^ 3 - x - 2
res <- itp(wiki, c(1, 2), k1 = 0.1, n0 = 1)

res1 <- print(res, all = FALSE)
test_that("print, all = FALSE", {
  testthat::expect_equal(res, res1)
})
res2 <- print(res, all = TRUE)
test_that("print, all = TRUE", {
  testthat::expect_equal(res, res2)
})


R/itp.R
#' The ITP root-finding algorithm
#'
#' Performs one-dimensional root-finding using the ITP algorithm of
#' Oliveira and Takahashi (2021).  The function \code{itp} searches an
#' interval [\eqn{a}, \eqn{b}] for a root (i.e., a zero) of the
#' function \code{f} with respect to its first argument. Each iteration
#' results in a bracketing interval for the root that is narrower than the
#' previous interval.  If the function is discontinuous then a point of
#' discontinuity at which the function changes sign may be found.
#'
#' @param f An R function or an external pointer to a C++ function.  For the
#'   latter see the article
#'   \href{https://gallery.rcpp.org/articles/passing-cpp-function-pointers/}{
#'   Passing user-supplied C++ functions} in the
#'   \href{https://gallery.rcpp.org/}{Rcpp Gallery}. The function for which the
#'   root is sought.
#' @param interval A numeric vector \code{c(a, b)} of length 2
#'   containing the end points of the interval to be searched for the root.
#'   The function values at the end points must be of opposite signs.
#' @param ... Additional arguments to be passed to \code{f}.
#' @param a,b An alternative way to set the lower and upper end points of the
#'   interval to be searched. The function values at these end points must be
#'   of opposite signs.
#' @param f.a,f.b The values of \code{f(a)} and \code{f(b)}, respectively.
#' @param epsilon A positive numeric scalar. The desired accuracy of the root.
#'   The algorithm continues until the width of the bracketing interval for the
#'   root is less than or equal to \code{2 * epsilon}.
#' @param k1,k2,n0 Numeric scalars. The values of the tuning parameters
#'   \ifelse{html}{\eqn{\kappa}\out{<sub>1</sub>}}{\eqn{\kappa_1}},
#'   \ifelse{html}{\eqn{\kappa}\out{<sub>2</sub>}}{\eqn{\kappa_2}},
#'   \ifelse{html}{\eqn{n}\out{<sub>0</sub>}}{\eqn{n_0}}.
#'   See \strong{Details}.
#' @details Page 8 of Oliveira and Takahashi (2021) describes the ITP
#'   algorithm and the roles of the tuning parameters
#'   \ifelse{html}{\eqn{\kappa}\out{<sub>1</sub>}}{\eqn{\kappa_1}},
#'   \ifelse{html}{\eqn{\kappa}\out{<sub>2</sub>}}{\eqn{\kappa_2}} and
#'   \ifelse{html}{\eqn{n}\out{<sub>0</sub>}}{\eqn{n_0}}.  The algorithm is
#'   described using pseudocode. The Wikipedia entry for the
#'   \href{https://en.wikipedia.org/wiki/ITP_method}{ITP method} provides
#'   a summary.  If the input function \code{f} is continuous over the interval
#'   [\code{a}, \code{b}] then the value of \code{f} evaluated at the estimated
#'   root is (approximately) equal to 0. If \code{f} is discontinuous over the
#'   interval [\code{a}, \code{b}] then the bracketing interval returned after
#'   convergence has the property that the signs of the function \code{f} at
#'   the end points of this interval are different and therefore the estimated
#'   root may be a point of discontinuity at which the sign of \code{f}
#'   changes.
#'
#'   The ITP method requires at most
#'   \ifelse{html}{\eqn{n}\out{<sub>max</sub>} = \eqn{n}\out{<sub>1/2</sub>} +
#'     \eqn{n}\out{<sub>0</sub>}}{\eqn{n_{\rm max} = n_{1/2} + n_0}} iterations,
#'   where \ifelse{html}{\eqn{n}\out{<sub>1/2</sub>}}{\eqn{n_{1/2}}} is the
#'   smallest integer not less than \ifelse{html}{log\out{<sub>2</sub>}((b-a) /
#'   2\eqn{\epsilon})}{\eqn{\log_2 ((b-a) / 2 \epsilon)}}.
#'   If \ifelse{html}{\eqn{n}\out{<sub>0</sub>} = 0}{\eqn{n_0 = 0}} then the
#'   ITP method will require no more iterations than the bisection method.
#'   Depending on the function \code{f}, setting a larger value for
#'   \ifelse{html}{\eqn{n}\out{<sub>0</sub>}}{\eqn{n_0 = 0}}, e.g. the default
#'   setting \ifelse{html}{\eqn{n}\out{<sub>0</sub>}=1}{\eqn{n_0 = 1}} used by
#'   the \code{itp} function, may result in a smaller number of iterations.
#'
#'   The default values of the other tuning parameters
#'   (\code{epsilon = 1e-10, k1 = 0.2 / (b - a), k2 = 2}) are set based on
#'   arguments made in Oliveira and Takahashi (2021).
#' @return An object (a list) of class \code{"itp"} containing the following
#'   components:
#'   \item{root}{the location of the root, calculated as \code{(a+b)/2}, where
#'     [\code{a, b}] is the bracketing interval after convergence.}
#'   \item{f.root}{the value of the function evaluated at root.}
#'   \item{iter}{the number of iterations performed.}
#'   \item{a,b}{the end points of the bracketing interval [\code{a, b}] after
#'     convergence.}
#'   \item{f.a,f.b}{the values of function at \code{a} and \code{b} after
#'     convergence.}
#'   \item{estim.prec}{an approximate estimated precision for \code{root},
#'     equal to the half the width of the final bracket for the root.}
#'     the root occurs at one of the input endpoints \code{a} or \code{b} then
#'     \code{iter = 0} and \code{estim.prec = NA}.
#'
#'   The returned object also has the attributes \code{f} (the input R function
#'   or pointer to a C++ function \code{f}), \code{f_args} (a list of
#'   additional arguments to \code{f} provided in \code{...}), \code{f_name}
#'   (a function name extracted from \code{as.character(substitute(f))} or the
#'   form of the R function if \code{f} was not named), \code{used_c} (a
#'   logical scalar: \code{FALSE}, if \code{f} is an R function and \code{TRUE}
#'   if \code{f} is a pointer to a C++ function) and \code{input_a} and
#'   \code{input_b} (the input values of \code{a} and \code{b}).  These
#'   attributes are used in \code{\link{plot.itp}} to produce a plot of the
#'   function \code{f} over the interval \code{(input_a, input_b)}.
#' @references Oliveira, I. F. D. and Takahashi, R. H. C. (2021). An Enhancement
#'   of the Bisection Method Average Performance Preserving Minmax Optimality,
#'   \emph{ACM Transactions on Mathematical Software}, \strong{47}(1), 1-24.
#'   \doi{10.1145/3423597}
#' @seealso \code{\link{print.itp}} and \code{\link{plot.itp}} for print and
#'   plot methods for objects of class \code{"itp"} returned from \code{itp}.
#' @examples
#' #### ----- The example used in the Wikipedia entry for the ITP method
#'
#' # Supplying an R function
#' wiki <- function(x) x ^ 3 - x - 2
#' itp(wiki, c(1, 2), epsilon = 0.0005, k1 = 0.1, n0 = 1)
#' # The default setting (with k1 = 0.2) wins by 1 iteration
#' wres <- itp(wiki, c(1, 2), epsilon = 0.0005, n0 = 1)
#' wres
#' plot(wres)
#'
#' # Supplying an external pointer to a C++ function
#' wiki_ptr <- xptr_create("wiki")
#' wres_c <- itp(f = wiki_ptr, c(1, 2), epsilon = 0.0005, k1 = 0.1)
#' wres_c
#' plot(wres_c)
#'
#' #### ----- Some examples from Table 1 of Oliveira and Takahashi (2021)
#'
#' ### Well-behaved functions
#'
#' # Lambert
#' lambert <- function(x) x * exp(x) - 1
#' itp(lambert, c(-1, 1))
#'
#' # Trigonometric 1
#' # Supplying an R function
#' trig1 <- function(x, root) tan(x - root)
#' itp(trig1, c(-1, 1), root = 1 / 10)
#' # Supplying an external pointer to a C++ function
#' trig1_ptr <- xptr_create("trig1")
#' itp(f = trig1_ptr, c(-1, 1), root = 1 / 10)
#'
#' # Logarithmic
#' logarithmic <- function(x, shift) log(abs(x - shift))
#' itp(logarithmic, c(-1, 1), shift = 10 /9)
#'
#' # Linear
#' linear <- function(x) x
#' # Solution in one iteration
#' itp(linear, c(-1, 1))
#' # Solution at an input endpoint
#' itp(linear, c(-1, 0))
#'
#' ### Ill-behaved functions
#'
#' ## Non-simple zero
#'
#' # Polynomial 3
#' poly3 <- function(x) (x * 1e6 - 1) ^ 3
#' itp(poly3, c(-1, 1))
#' # Using n0 = 0 leads to fewer iterations, in this example
#' poly3 <- function(x) (x * 1e6 - 1) ^ 3
#' itp(poly3, c(-1, 1), n0 = 0)
#'
#' ## Discontinuous
#'
#' # Staircase
#' staircase <- function(x) ceiling(10 * x - 1) + 1 / 2
#' itp(staircase, c(-1, 1))
#'
#' ## Multiple roots
#'
#' # Warsaw
#' warsaw <- function(x) ifelse(x > -1, sin(1 / (x + 1)), -1)
#' # Function increasing over the interval
#' itp(warsaw, c(-1, 1))
#' # Function decreasing over the interval
#' itp(warsaw, c(-0.85, -0.8))
#' @export
itp <- function(f, interval, ..., a = min(interval), b = max(interval),
                f.a = f(a, ...), f.b = f(b, ...), epsilon = 1e-10,
                k1 = 0.2 / (b - a), k2 = 2, n0 = 1) {
  if (is.function(f)) {
    using_c <- FALSE
  } else if (inherits(f, "externalptr")) {
    using_c <- TRUE
  } else {
    stop("''f'' must be an R function or a pointer to a C++ function")
  }
  if (!missing(interval) && length(interval) != 2L) {
    stop("'interval' must be a vector of length 2")
  }
  if (!is.numeric(a) || !is.numeric(b) || a >= b) {
    stop("a < b  is not fulfilled")
  }
  if (k1 <= 0) {
    stop("k1 must be positive")
  }
  if (k2 < 1 || k2 >= 1 + (1 + sqrt(5)) / 2) {
    stop("k2 must be in [ 1, 1 + (1 + sqrt(5)) / 2 )")
  }
  if (n0 < 0) {
    stop("n0 must be non-negative")
  }
  # Save the input values of a and b for use in plot.itp
  input_a <- a
  input_b <- b
  # If we are using C++ then calculate f(a) and f(b)
  if (using_c) {
    f.a <- do.call(xptr_eval, list(a, list(...), f))
    f.b <- do.call(xptr_eval, list(b, list(...), f))
  }
  # Check that f(a) and f(b) are finite
  if (!all(is.finite(c(f.a, f.b)))) {
    stop("f(a) and f(b) must be finite")
  }
  # Create a function name
  temp <- as.character(substitute(f))
  if (using_c) {
    f_name <- temp
  } else {
    f_name <- ifelse(length(temp) > 1, temp[3], temp)
  }
  # Check whether (a, b) already satisfies the convergence criterion
  if (b - a <= 2 * epsilon) {
    root <- (a + b) / 2
    val <- list(root = root, f.root = f(root, ...), iter = 0, a = a,
                b = b, f.a = f.a, f.b = f.b, estim.prec = (b - a) / 2)
    attributes(val) <- c(attributes(val), list(f = f, f_args = list(...),
                         input_a = input_a, input_b = input_b, f_name = f_name,
                         used_c = using_c))
    class(val) <- "itp"
    return(val)
  }
  # Check whether the root lies on a limit of the input interval
  if (f.a == 0) {
    val <- list(root = a, f.root = 0, iter = 0, a = a,
                b = b, f.a = f.a, f.b = f.b, estim.prec = NA)
    attributes(val) <- c(attributes(val), list(f = f, f_args = list(...),
                         input_a = input_a, input_b = input_b, f_name = f_name,
                         used_c = using_c))
    class(val) <- "itp"
    return(val)
  } else if (f.b == 0) {
    val <- list(root = b, f.root = 0, iter = 0, a = a,
                b = b, f.a = f.a, f.b = f.b, estim.prec = NA)
    attributes(val) <- c(attributes(val), list(f = f, f_args = list(...),
                         input_a = input_a, input_b = input_b, f_name = f_name,
                         used_c = using_c))
    class(val) <- "itp"
    return(val)
  }
  # Check that f(a) and f(b) have opposite signs
  if (sign(f.a) * sign(f.b) > 0) {
    stop("the f() values at the end points are not of opposite sign")
  }
  log2e <- log2(epsilon)
  log2bma <- log2(b - a)
  # for_rk = epsilon * 2 ^ (n12 + n0)
  # log(for_rk) = log2(epsilon) + (n12 + n0)
  # n_1/2 (in eqn (3)) = ceiling(log2bma - log2e) - 1
  for_rk <- 2 ^ (n0 - 1 + log2e + ceiling(log2bma - log2e))
  # Call itp_r() or itp_cpp() as appropriate
  if (using_c) {
    for_itp_cpp <- list(f = f, pars = list(...), a = a, b = b, ya = f.a,
                        yb = f.b, epsilon = epsilon, k1 = k1, k2 = k2,
                        for_rk = for_rk, inc = sign(f.b))
    val <- do.call("itp_cpp", for_itp_cpp)
  } else {
    val <- itp_r(f = f, ..., a = a, b = b, ya = f.a, yb = f.b,
                 epsilon = epsilon, k1 = k1, k2 = k2, for_rk = for_rk,
                 inc = sign(f.b))
  }
  attributes(val) <- c(attributes(val), list(f = f, f_args = list(...),
                       input_a = input_a, input_b = input_b, f_name = f_name,
                       used_c = using_c))
  class(val) <- "itp"
  return(val)
}


R/itpPlot.R
#' Plot method for objects of class \code{"itp"}
#'
#' Plot method for objects of class \code{"itp"} returned from
#' \code{\link{itp}}.
#' @param x An object inheriting from class \code{"itp"}, a result of a
#'   call to \code{\link{itp}}.
#' @param ... Arguments passed to \code{\link[graphics]{curve}},
#'   such as graphical parameters.
#' @details Uses \code{\link[graphics]{curve}} to produce a plot of the
#'   function \code{f} provided to \code{\link{itp}} over the interval within
#'   which a root was sought. The estimated root is indicated using a
#'   horizontal line drawn at 0 and a vertical line drawn at the estimated
#'   root. By default the name of the function \code{f} is used as a title,
#'   but this can be replaced by supplying the argument \code{main}.  The
#'   interval over which \code{f} is plotted can be changed by supplying the
#'   arguments \code{from} and/or \code{to}.
#' @return No return value, only the plot is produced.
#' @examples
#' # Lambert
#'
#' # Supplying an R function
#' lambert <- function(x) x * exp(x) - 1
#' x <- itp(lambert, c(-1, 1))
#' plot(x)
#'
#' # Supplying an external pointer to a C++ function
#' lambert_ptr <- xptr_create("lambert")
#' x <- itp(lambert_ptr, c(-1, 1))
#' plot(x, main = "Lambert")
#' @seealso \code{\link{itp}} for the Interpolate, Truncate, Project (ITP) root
#'   finding algorithm.
#' @export
plot.itp <- function(x, ...) {
  # Extract f and any additional arguments to f
  f <- attr(x, "f")
  f_args <- attr(x, "f_args")
  # If f is a C++ function
  if (attr(x, "used_c")) {
    plotfun <- function(x) {
      not_vectorised <- function(x) {
        do.call(xptr_eval, list(x, f_args, f))
      }
      return(vapply(x, not_vectorised, 0.0))
    }
  } else {
    if (length(f_args) == 0) {
      plotfun <- f
    } else {
      plotfun <- function(x) {
        return(do.call(f, c(list(x = x), f_args)))
      }
    }
  }
  curvefn <- function(x, ..., from = attr(x, "input_a"),
                      to = attr(x, "input_b"), xlab = "x", ylab = "f(x)",
                      main = attr(x, "f_name")) {
    graphics::curve(plotfun, ..., from = from, to = to, xlab = xlab,
                    ylab = ylab, main = main)
    graphics::abline(h = 0, lty = 2)
    graphics::abline(v = x$root, lty = 2)
  }
  curvefn(x, ...)
  return(invisible())
}


R/itpPrint.R
#' Print method for objects of class \code{"itp"}
#'
#' Prints objects of class \code{"itp"} returned from \code{\link{itp}}.
#' @param x An object inheriting from class \code{"itp"}, a result of a
#'   call to \code{\link{itp}}.
#' @param all A logical scalar.  If \code{all = FALSE} then only the estimated
#'   root, the value of the function at the root an the number of iterations
#'   are printed.  If \code{all = TRUE} then, in addition, the final bracketing
#'   interval [\code{a, b}], the values of the function at the end points of
#'   this interval and the estimated precision are printed.
#' @param digits The argument \code{digits} passed to \code{\link{format}}
#'   to set the number of significant digits to print.
#' @param ... Further arguments to be passed to or from other methods. They are
#'   ignored in this function..
#' @details The default setting is to print only the root, the value of the
#'   function at the root and the number of iterations.  To include the
#'   bracketing interval after convergence and the estimated precision use
#'   \code{all = TRUE}.
#'
#' @return The argument \code{x} is returned, invisibly.
#' @seealso \code{\link{itp}} for the Interpolate, Truncate, Project (ITP) root
#'   finding algorithm.
#' @export
print.itp <- function(x, all = FALSE,
                      digits = max(3L, getOption("digits") - 3L), ...) {
  temp <- x
  names(temp) <- c("root", "f(root)", "iterations", "a", "b", "f.a", "f.b",
                   "precision")
  cat("function:", attr(x, "f_name"), "\n")
  if (!all) {
    print.default(format(temp[1:3], digits = digits), print.gap = 2L,
                  quote = FALSE)
  } else {
    print.default(format(temp[1:8], digits = digits), print.gap = 2L,
                  quote = FALSE)
  }
  invisible(x)
}


R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @keywords internal
#' @rdname itp-internal
itp_cpp <- function(f, pars, a, b, ya, yb, epsilon, k1, k2, for_rk, inc) {
    .Call(`_itp_itp_cpp`, f, pars, a, b, ya, yb, epsilon, k1, k2, for_rk, inc)
}

#' Call a C++ function using an external pointer
#'
#' This function is used in \code{\link{plot.itp}} to plot a function and
#' the root estimated by \code{\link{itp}}.
#' @param x The main argument of the function.
#' @param pars A list of additional arguments to the function.  This may be an
#'   empty list.
#' @param xpsexp An external pointer to a C++ function.
#' @details See the
#' \href{https://gallery.rcpp.org/articles/passing-cpp-function-pointers/}{
#' Passing user-supplied C++ functions} article in the
#' \href{https://gallery.rcpp.org/}{Rcpp Gallery} for information.
#' @return A numeric scalar: the value of the C++ function evaluated at the
#'   input values \code{x} and \code{pars}.
#' @examples
#' lambert_ptr <- xptr_create("lambert")
#' res <- itp(lambert_ptr, c(-1, 1))
#'
#' # Value at lower limit
#' xptr_eval(-1, list(), lambert_ptr)
#'
#' # Value at upper limit
#' xptr_eval(1, list(), lambert_ptr)
#'
#' # Value at the estimated root
#' xptr_eval(res$root, list(), lambert_ptr)
#' @seealso \code{\link{xptr_create}} for creating an external pointer to a
#'   C++ function.
#' @export
xptr_eval <- function(x, pars, xpsexp) {
    .Call(`_itp_xptr_eval`, x, pars, xpsexp)
}

#' The ITP root-finding algorithm using C++
#'
#' Performs one-dimensional root-finding using the ITP algorithm of
#' Oliveira and Takahashi (2021). This function is equivalent to
#' \code{\link{itp}} but calculations are performed entirely using C++, and
#' the arguments differ slightly: \code{itp_c} has a named required argument
#' \code{pars} rather than \code{...} and it does not have the arguments
#' \code{interval}, \code{f.a} or \code{f.b}.
#' @param f An R function or an external pointer to a C++ function.  For the
#'   latter see the article
#'   \href{https://gallery.rcpp.org/articles/passing-cpp-function-pointers/}{
#'   Passing user-supplied C++ functions} in the
#'   \href{https://gallery.rcpp.org/}{Rcpp Gallery}. The function for which the
#'   root is sought.
#' @param f An external pointer to a C++ function that evaluates the function
#'   \eqn{f}.
#' @param pars A list of additional arguments to the function.  This may be an
#'   empty list.
#' @param a,b Numeric scalars. Lower (\code{a}) and upper \code{b} limits of
#'   the interval to be searched for a root.
#' @param epsilon A positive numeric scalar. The desired accuracy of the root.
#'   The algorithm continues until the width of the bracketing interval for the
#'   root is less than or equal to \code{2 * epsilon}.
#' @param k1,k2,n0 Numeric scalars. The values of the tuning parameters
#'   \ifelse{html}{\eqn{\kappa}\out{<sub>1</sub>}}{\eqn{\kappa_1}},
#'   \ifelse{html}{\eqn{\kappa}\out{<sub>2</sub>}}{\eqn{\kappa_2}},
#'   \ifelse{html}{\eqn{n}\out{<sub>0</sub>}}{\eqn{n_0}}.
#'   See the \strong{Details} section of \code{\link{itp}}.
#'
#'   The default value for \code{k1} in \code{\link{itp_c}} is set as
#'   the inadmissible value of \code{-1}
#'   (in reality \ifelse{html}{\eqn{\kappa}\out{<sub>1</sub>}}{\eqn{\kappa_1}}
#'   must be positive) as a device to set the same default value for \code{k1}
#'   as \code{\link{itp}}, that is, \code{k1 = 0.2 / (b - a)}.  If the input
#'   value of \code{k1} is less than or equal to 0 then, inside
#'   \code{\link{itp_c}}, \code{k1 = 0.2 / (b - a)} is set.
#' @details For details see \code{\link{itp}}.
#' @return An object (a list) of class \code{"itp"} with the same structure
#'   as detailed in the \strong{Value} section of \code{\link{itp}}, except
#'   that the attribute \code{f_name} is empty (equal to \code{""}).
#' @references Oliveira, I. F. D. and Takahashi, R. H. C. (2021). An Enhancement
#'   of the Bisection Method Average Performance Preserving Minmax Optimality,
#'   \emph{ACM Transactions on Mathematical Software}, \strong{47}(1), 1-24.
#'   \doi{10.1145/3423597}
#' @seealso \code{\link{print.itp}} and \code{\link{plot.itp}} for print and
#'   plot methods for objects of class \code{"itp"} returned from \code{itp_c}
#'   or \code{itp}.
#' @examples
#' wiki_ptr <- xptr_create("wiki")
#' wres <- itp_c(f = wiki_ptr, pars = list(), a = 1, b = 2, epsilon = 0.0005)
#' wres
#' plot(wres, main = "Wiki")
#' @export
itp_c <- function(f, pars, a, b, epsilon = 1e-10, k1 = -1.0, k2 = 2.0, n0 = 1.0) {
    .Call(`_itp_itp_c`, f, pars, a, b, epsilon, k1, k2, n0)
}

wiki_cpp <- function(x, pars) {
    .Call(`_itp_wiki_cpp`, x, pars)
}

neg_wiki_cpp <- function(x, pars) {
    .Call(`_itp_neg_wiki_cpp`, x, pars)
}

lambert_cpp <- function(x, pars) {
    .Call(`_itp_lambert_cpp`, x, pars)
}

trig1_cpp <- function(x, pars) {
    .Call(`_itp_trig1_cpp`, x, pars)
}

poly3_cpp <- function(x, pars) {
    .Call(`_itp_poly3_cpp`, x, pars)
}

linear_cpp <- function(x, pars) {
    .Call(`_itp_linear_cpp`, x, pars)
}

staircase_cpp <- function(x, pars) {
    .Call(`_itp_staircase_cpp`, x, pars)
}

log_cpp <- function(x, pars) {
    .Call(`_itp_log_cpp`, x, pars)
}

#' Create an external pointer to a C++ function
#'
#' This function is used in the \code{\link[itp:itp-package]{itp}} package to
#' create external pointers to the C++ functions used as examples to
#' illustrate the use of the function \code{\link{itp}}.  These pointers are
#' passed as the argument \code{f} to \code{\link{itp}}.  To create their own
#' examples the user will need to create their own C++ function(s) and a
#' function that is similar to \code{xptr_create}.
#'
#' @param fstr A string indicating the C++ function required.
#' @details See the vignette
#' \href{https://paulnorthrop.github.io/itp/articles/itp-vignette.html}{
#' Overview of the itp package} and the file
#' \href{https://raw.githubusercontent.com/paulnorthrop/itp/main/src/user_fns.cpp}{
#' user_fns.cpp} for information.
#'
#' The example C++ functions available in \code{itp} are: \code{"wiki"},
#' \code{"lambert"}, \code{"trig1"}, \code{"poly3"}, \code{"linear"},
#' \code{"warsaw"} and \code{staircase}.
#' @return The external pointer.
#' @seealso \code{\link{xptr_eval}} for calling a C++ function using an
#'   external pointer.
#' @examples
#' lambert_ptr <- xptr_create("lambert")
#' res <- itp(lambert_ptr, c(-1, 1))
#' @export
xptr_create <- function(fstr) {
    .Call(`_itp_xptr_create`, fstr)
}

# Register entry points for exported C++ functions
methods::setLoadAction(function(ns) {
    .Call(`_itp_RcppExport_registerCCallable`)
})
